test/gtest/uct/ib/test_devx.cc

```diff
@@ -49,3 +51,216 @@ UCS_TEST_P(test_devx, dbrec)
 }
 
 UCT_INSTANTIATE_IB_TEST_CASE(test_devx);
+
+
+class test_devx_umr_mkey : public test_devx {
+public:
+    void init() {
+        test_devx::init();
+    }
+
+    bool check_xgvmi() const {
+        return md()->flags & UCT_IB_MLX5_MD_FLAG_INDIRECT_XGVMI;
+    }
+
+    /*
+     * XGVMI capability advertised by device might be dropped on the initial
+     * operation. This function performs dummy XGVMI operation to detect
+     * whether XGVMI is actually operational
+     */
+    bool check_xgvmi_effective() const {
+        if (check_xgvmi()) {
+            uct_ib_mlx5_devx_mem_t *memh = create_memh(1);
+            ASSERT_UCS_OK(uct_ib_mlx5_devx_reg_exported_key(md(), memh));
+            /* XGVMI capability might be dropped by this point! */
+            destroy_memh(memh);
+            return check_xgvmi();
+        }
+
+        return false;
+    }
+
+    uct_ib_mlx5_devx_mem_t *create_memh(size_t length) const {
+        void *addr                     = ucs_malloc(length, "test");
+        uct_md_mem_reg_params_t params = {};
+        uct_mem_h memh;
+
+        EXPECT_NE(nullptr, addr);
+        ASSERT_UCS_OK(uct_ib_mlx5_devx_mem_reg(m_e->md(), addr, length, &params,
+                                               &memh));
+        return ucs_derived_of(memh, uct_ib_mlx5_devx_mem_t);
+    }
+
+    void destroy_memh(uct_ib_mlx5_devx_mem_t *memh) const {
+        void *addr                       = memh->address;
+        uct_md_mem_dereg_params_t params = {};
+
+        params.field_mask = UCT_MD_MEM_DEREG_FIELD_MEMH;
+        params.memh       = (uct_mem_h)memh;
+
+        ASSERT_UCS_OK(uct_ib_mlx5_devx_mem_dereg(m_e->md(), &params));
+
+        if (nullptr != addr) {
+            ucs_free(addr);
+        }
+    }
+
+    void check_umr_init(bool init_export, bool init_import) const {
+        EXPECT_EQ(init_export, nullptr != md()->umr.qp);
```

Comment:
if (init_export) {
  EXPECT_NE(nullptr, umr.qp);
  EXPECT_NE(nullptr, umr.cq);
} else {
   EXPECT_TRUE(ucs_list_is_empty(&md()->umr.mkey_pool));
}

Reply:
i think it's not important to check if umr objects are initialized, but just try to use them (it should fail if they are not initialized)

Reply:
well, the most important use case here is to check that UMR objects are NOT initialized (when they should not be)